packages/nodejs/src/span_processor.ts

```diff
@@ -22,22 +21,18 @@ export class SpanProcessor implements OpenTelemetrySpanProcessor {
   onStart(_span: Span, _parentContext: Context): void {}
 
   onEnd(span: ReadableSpan): void {
-    const appsignalSpan = this.client.tracer().currentSpan()
-
-    if (!(appsignalSpan instanceof NoopSpan)) {
-      this.client.extension.importOpenTelemetrySpan(
-        span.spanContext().spanId,
-        appsignalSpan.spanId,
-        appsignalSpan.traceId,
-        span.startTime[0],
-        span.startTime[1],
-        span.endTime[0],
-        span.endTime[1],
-        span.name,
-        span.attributes,
-        span.instrumentationLibrary.name
-      )
-    }
+    this.client.extension.importOpenTelemetrySpan(
+      span.spanContext().spanId,
+      span.parentSpanId || "",
+      span.spanContext().traceId,
+      span.startTime[0],
+      span.startTime[1],
+      span.endTime[0],
+      span.endTime[1],
+      span.name,
+      span.attributes,
+      span.instrumentationLibrary.name
+    )
```

Comment:
If I do this, I get empty span and trace ids. For some reason we create 3-4 spans on boot (looks like outgoing requests 🤨 ) and it gets stuck on that root span so it keeps using a root span that's been completed already. It makes very little sense as per usual.

```diff
diff --git packages/nodejs/src/span_processor.ts packages/nodejs/src/span_processor.ts
index 2d7c247..a1d764a 100644
--- packages/nodejs/src/span_processor.ts
+++ packages/nodejs/src/span_processor.ts
@@ -5,6 +5,7 @@ import {
   SpanProcessor as OpenTelemetrySpanProcessor
 } from "./interfaces/span_processor"
 import { BaseClient as Client } from "./client"
+import { NoopSpan } from "./noops"

 export class SpanProcessor implements OpenTelemetrySpanProcessor {
   client: Client
@@ -21,10 +22,23 @@ export class SpanProcessor implements OpenTelemetrySpanProcessor {
   onStart(_span: Span, _parentContext: Context): void {}

   onEnd(span: ReadableSpan): void {
+    const appsignalSpan = this.client.tracer().currentSpan()
+
+    let parentSpanId
+    let traceId
+    if (!(appsignalSpan instanceof NoopSpan)) {
+      parentSpanId = appsignalSpan.spanId
+      traceId = appsignalSpan.traceId
+      console.log("!!! appsignalSpan", appsignalSpan, { parentSpanId, traceId })
+    } else {
+      parentSpanId = span.parentSpanId
+      traceId = span.spanContext().traceId
+      console.log("!!! opentelemetrySpan", span, { parentSpanId, traceId })
+    }
     this.client.extension.importOpenTelemetrySpan(
       span.spanContext().spanId,
-      span.parentSpanId || "",
-      span.spanContext().traceId,
+      parentSpanId || "",
+      traceId,
       span.startTime[0],
       span.startTime[1],
```

Reply:
Oof, I see. Looks like a different variation of [this issue](https://github.com/appsignal/appsignal-nodejs/pull/656)... ignore this then. Anyway, thanks for trying it out :)

Reply:
I think so yes. Because I see 3 spans being created and 4 being closed (yes those numbers do not match). Whatever should unregister the root span (if anything) doesn't. The main issue is that `currentSpan` is not very reliable to return an active, not completed, span.

Which we can also dive into to fix, but we can only resolve so many issues in something we're going to deprecate and remove eventually anyway. How important is the compatibility layer in this transition period?

Reply:
I've attempted to make a fix in https://github.com/appsignal/appsignal-nodejs/pull/672

Reply:
I've done more testing. What will work now with PR #672, is either using AppSignal or OpenTelemetry. People are not forced to switch over if they want to continue using the AppSignal express and/or HTTP instrumentations.

However, you can't mix both instrumentations, because the OpenTelemetry HTTP and express instrumentations overwrites/breaks the AppSignal HTTP and express instrumentations.

Reply:
That makes sense! I can't think of an use-case where using OpenTelemetry HTTP and AppSignal HTTP simultaneously, and receiving spans for both, is a meaningful thing that someone would want.

What I think would be nice to have is, say, combinations like OpenTelemetry HTTP+NestJS with AppSignal Redis, or AppSignal HTTP+Express with OpenTelemetry MySQL. If #672 supports those use cases, then it's perfect.

Reply:
It will only support either AppSignal root spans with OpenTelemetry child spans, or OpenTelemetry root spans with OpenTelemetry child spans. You can't use OpenTelemetry root spans with AppSignal child spans I think, because we don't check in the AppSignal tracer/scope/span if a root OpenTelemetry span is available. This PR is only to improve the OpenTelemetry spanprocessor.

Reply:
Ah, I see. That makes sense, of course, given the implementation. I think this is great. It provides a way for users to begin using OpenTelemetry for the instrumentation-specific thing they care about (MySQL, IORedis) without having to change the rest of their stack. 👍